extmod/vfs_rom.c

```diff
@@ -0,0 +1,398 @@
+// can be memory mapped.
+//
+// Concepts:
+// - varuint: An unsigned integer that is encoded in a variable number of bytes. It is
+//   stored big-endian with the high bit of the byte set if there are following bytes.
+// - record: A variable sized element with a type.  It is stored as two varuint's and then
+//   a payload.  The first varuint is the record kind and the second varuint is the
+//   payload length (which may be zero bytes long).
+//
+// A ROMFS filesystem is a record with record kind 0x14a6b1, chosen so the encoded value
+// is 0xd2-0xcd-0x31 which is "RM1" with the first two bytes having their high bit set.
+// If the ROMFS record's payload is non-empty then it contains records.
+//
+// Record types:
+// - 0 = unused, can be used to detect corruption of the filesystem.
+// - 1 = padding/comments, can contain any data in their payload.
+// - 2 = verbatim data, used to store file data.
+// - 3 = indirect data, pointer to offset within the ROMFS payload.
+// - 4 = a directory: payload contains a varuint which is the length of the directory
+//       name in bytes, then the name, then optional nested records.
+// - 5 = a file: payload contains a varuint which is the length of the filename in bytes
+//       then the name, then optional nested records.
+//
+// Remarks:
+// - A varuint can be padded if needed by prepending with one or more 0x80 bytes.  This
+//   padding does not change any semantics.
+// - The size of the ROMFS record (including kind and length and payload) must be a
+//   multiple of 2 (because it's not possible to add a padding record of one byte).
+// - File data can be optionally aligned using padding records and/or indirect data
+//   records.
+// - There is no limit to the size of directory/file names or file data.
+//
+// Unknown record types must be skipped over.  They may in the future add optional
+// features, while still retaining backwards compatibility.  Such features may be:
+// - Alignment requirements of the ROMFS record.
+// - Timestamps on directories/files.
+// - A precomputed hash of a file, or other metadata.
+// - An optimised lookup table indexing the directory hierarchy.
+
+#include <string.h>
+
+#include "py/bc.h"
+#include "py/runtime.h"
+#include "py/mperrno.h"
+#include "extmod/vfs.h"
+#include "extmod/vfs_rom.h"
+
+#if MICROPY_VFS_ROM
+
+#define ROMFS_SIZE_MIN (4)
+#define ROMFS_HEADER_BYTE0 (0x80 | 'R')
+#define ROMFS_HEADER_BYTE1 (0x80 | 'M')
+#define ROMFS_HEADER_BYTE2 (0x00 | '1')
+
+enum {
+    ROMFS_RECORD_KIND_UNUSED = 0,
+    ROMFS_RECORD_KIND_PADDING,
+    ROMFS_RECORD_KIND_DATA_VERBATIM,
+    ROMFS_RECORD_KIND_DATA_POINTER,
+    ROMFS_RECORD_KIND_DIRECTORY,
+    ROMFS_RECORD_KIND_FILE,
+};
+
+struct _mp_obj_vfs_rom_t {
+    mp_obj_base_t base;
+    mp_obj_t memory;
+    const uint8_t *filesystem;
+    const uint8_t *filesystem_top;
+};
+
+static mp_uint_t extract_record(const uint8_t **fs, const uint8_t **fs_next) {
```

Comment:
I'm not sure it's possible to use the enum, because the FS should support arbitrary record kinds, eg they might be 32-bit numbers.  If the C compiler sees that it's an enum with only a handful of possibilities, it may narrow the enum type to 8 bits, and hence a large and unknown record type might be truncated to 8 bits and alias a known one.

Instead, maybe it's better to change these constants to simple #define's, and then just have a typedef for `record_kind_t` to be `mp_uint_t`.

Reply:
I added a typedef for `record_kind_t` and changed the enum to #define's (because it's part of the filesystem spec, IMO better to explicitly list the values for the records).

Reply:
For once our old friend the C standard has good news! C99 6.4.4.3 Enumeration constants: "An identifier declared as an enumeration constant has type int." The storage size should be the same as mp_uint_t.

There *is* a gcc option, `-fshort-enums` that allows the non-standard behaviour you're describing. I just noticed that it is currently turned on for the nrf port (only), but I suggest we disable it there rather than accommodate it.

Reply:
> Our replies crossed, but that seems fine to me too. 😄

Haha, OK!

If you think it's fine, let's leave it as a `typedef mp_uint_t record_kind_t`.

Reply:
> I don't know how easy it is for us to warn against using it.

`MP_STATIC_ASSERT(sizeof(some_enum) == sizeof(int))` would catch it, but I understand your reasoning. No worries.